### ISSUE TITLE
devimint docker container

### ISSUE THREAD

**Issue by justinmoon** [fedimint/fedimint#3061]
Some frontend developers working on https://github.com/fedimint/ui prefer to use docker to nix since it's a more familiar tool. 

So that repo is currently mainaining a nix config and a `docker-compose.yml`. Unfortunately, this is a little error-prone. Here are 2 issues surfaced today:

- https://github.com/fedimint/ui/pull/161
- https://github.com/fedimint/ui/pull/158

Perhaps it would be better to package devimint as a docker container? This way the the backend would be run in the same way whether the dev used docker or nix directly, but frontend devs wouldn't have to install nix at all.

Not attached to this idea, feel free to shoot holes in it.

The upside of having fedimint/ui maintain it's own `docker-compose.yml` is that it's the only thing consuming the docker containers right now. But hopefully this changes soon.

**Comment by elsirion** [comment]
Shouldn't it be easy to build a container off our nix env? I don't want docker to become a required part of my dev setup but would be happy to just have a nix output that builds+publishes a devimint container.

**Comment by justinmoon** [comment]
> would be happy to just have a nix output that builds+publishes a devimint container.

Yes, this is the goal. An option to run devimint via docker for those who don't want nix to be part of their dev environment.

**Comment by bradleystachurski** [comment]
I haven't worked with the UI yet, so I'm not familiar with the local dev flow, but I've been using a hybrid docker/nix approach to keep the project sandboxed on my dev box. Would these containers be helpful? If so, I can merge into a single Dockerfile and add the required files in a UI PR. I can also investigate building/publishing a devimint container as a nix output.

```bash
» cat nix/Dockerfile
FROM ubuntu:20.04

RUN apt-get update && apt-get install -y sudo

# Disable password for sudo users
RUN echo '%sudo ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers

RUN addgroup --system nixbld && \
  adduser --home /home/nix --disabled-password --gecos "" --shell /bin/bash nix && \
  adduser nix nixbld && \
  adduser nix sudo && \
  mkdir -m 0755 /nix && chown nix /nix && \
  mkdir -p /etc/nix && \
  echo 'sandbox = false' > /etc/nix/nix.conf && \
  echo 'experimental-features = nix-command flakes' >> /etc/nix/nix.conf

CMD /bin/bash -l
USER nix
ENV USER nix
WORKDIR /home/nix

RUN sudo apt-get update && sudo apt-get install -y \
	curl \
	git \
	xz-utils \
	vim

RUN touch .bash_profile && \
  curl -L https://nixos.org/nix/install | sh

RUN echo ". /home/nix/.nix-profile/etc/profile.d/nix.sh" >> ~/.bashrc

ENTRYPOINT /bin/bash
```
```bash
» cat fedimint/Dockerfile
FROM nix-dev

RUN sudo apt-get update && sudo apt-get install -y git colorized-logs

RUN git clone https://github.com/fedimint/fedimint.git
WORKDIR /home/nix/fedimint

# trusted settings will prevent being prompted with (y/N) questions when running nix build
COPY trusted-settings.json /home/nix/.local/share/nix/trusted-settings.json
RUN sudo chown -R nix:nix /home/nix/.local

# update /nix/store so `nix develop` is fast in the container
RUN /home/nix/.nix-profile/bin/nix develop
```

**Comment by bradleystachurski** [comment]
Potential can of worms, so feel free to ignore this comment 😃 

These docs (https://github.com/fedimint/ui/pull/158/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R64) are suggesting a quasi-monorepo. Was there a motivation for not choosing a monorepo that includes both projects?

**Comment by EthnTuttle** [comment]
> Potential can of worms, so feel free to ignore this comment 😃
> 
> These docs (https://github.com/fedimint/ui/pull/158/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R64) are suggesting a quasi-monorepo. Was there a motivation for not choosing a monorepo that includes both projects?

The UI code used to live in the [main fedimint repo](https://github.com/fedimint/fedimint/pull/2733). While the UI repo is a good example and default for a fedimint admin and gateway admin page, it didn't need to be tightly coupled to the protocol repo since there's is an anticipation that people will build out their own apps for admin/management of the federation.

Also, there was a lot of noise generated with UI/UX issues that cluttered up the main fedimint GH. By having it's own repo, design and UI folks could iterate in parallel. 

Also also, the two dev deps are very different. you shouldn't need 40GB+ of nix to develop on the UI. By removing nix and making docker the dev env of choice, it lowered the barrier to entry for the general web dev crowd.

I may be forgetting a reason or two and there is probably a way to make it work as a mono-repo. 🤷‍♂️

>  I can also investigate building/publishing a devimint container as a nix output.

I personally would love to see this as I could see devimint becoming an industry wide tool for bitcoin dev.

**Comment by bradleystachurski** [comment]
@EthnTuttle Thanks for the context!

Let's touch base during today's dev call about a plan that works well with folks working on the UI.

**Comment by justinmoon** [comment]
Even if we went with a monorepo, I still think a devimint docker container would be useful because it would allow frontend devs to do their thing without needing to install nix or compile rust packages

**Comment by dpc** [comment]
This issue exposes one of the fundamental problems with Docker: each image starts a separate "container" (effectively: machine). This is OK for stuff that is at the core a network exposed API, but not for tools you want to use "inline".


One workaround is to run with `host` network: https://docs.docker.com/compose/compose-file/compose-file-v3/#network_mode , but then there's still problem how are you going to be calling binaries from the docker image itself, and glue it all together, and I'm not sure how well it works on MacOS (where VMs are used for container, AFAIU).

Another one would be to use the portable binaries that Fedimint's CI produces somehow.

There might be some other approaches how to get it done. I've seen so many projects inventing complex variations of solutions for this problem (like using docker volume binding to mount source code into a container in which they actually build and run stuff) etc.

IMO the right one is to recognize Nix the right true way, accept Flake into your life, study Nix pills and then go tell everyone you know about it. :D
